doc/ldaptree.1.in
.\"
.\" LDAP Utilities
.\" Copyright (C) 2019 David M. Syzdek <david@syzdek.net>.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"    1. Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"    2. Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in the
.\"       documentation and/or other materials provided with the distribution.
.\"
.\"    3. Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived from
.\"       this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
.\" IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
.\" THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" doc/ldap2csv.1.in - man page for ldap2csv
.\"
.TH "LDAPTREE" "1" "@RELEASE_MONTH@" "@PACKAGE_NAME@" "User Commands"
.SH NAME
ldaptree \- LDAP search tool which displays results in tree form


.SH SYNOPSIS
\fBldaptree\fR
[\fB-b\fR \fIbasedn\fR]
[\fB-c\fR]
[\fB-d\fR \fIlevel\fR]
[\fB-D\fR \fIbinddn\fR]
[\fB-H\fR \fIURI\fR]
[\fB-l\fR \fIlimit\fR]
[\fB-L\fR[\fB-L\fR]]
[\fB-n\fR]
[\fB-v\fR | \fB--version\fR]
[\fB-s\fR \fIscope\fR]
[\fB-S\fR \fIattr\fR]
[\fB-w\fR \fIpasswd\fR]
[\fB-W\fR]
[\fB-x\fR]
[\fB-y\fR \fIfile\fR]
[\fB-Y\fR \fImech\fR]
[\fB-z\fR \fIlimit\fR]
[\fB-Z\fR[\fB-Z\fR]]
[\fIfilter\fR]
[\fIattributes...\fR]
.sp
\fBldaptree\fR [ \fB-h\fR | \fB--help\fR ]
.sp
\fBldaptree\fR [ \fB-V\fR | \fB--version\fR ]

.SH DESCRIPTION
\fBldaptree\fR is a shell utilty which performs an LDAP search and prints the
results as either an ASCII graph or a bulleted list. By default the utilty
does not retrieve any attributes from LDAP servers, however if a \fIfilter\fR
and \fIattributes\fR are provided as command line arguments, then the
attribute values will be displayed inline with the results.

.SH OPTIONS
.TP
\fB-c\fR
do not stop if an error is encountered
.TP
\fB-d\fR
set OpenLDAP debug level to `level'
.TP
\fB-D\fR \fIbinddn\fR
bind DN used for simple bind
.TP
\fB-H\fR \fIURI\fR
specifies list of LDAP Uniform Resource Identifier(s) used to connect to LDAP server
.TP
\fB-l\fR \fIlimit\fR
time limit (in seconds) for search
.TP
\fB-L\fR
two \fB-L\fR disables comments
.TP
\fB-v\fR   \fB--version\fR
run in verbose mode
.TP
\fB-s\fR \fIscope\fR
specifies search filter. Must be one of \fIbase\fR, \fIone\fR, \fIsub\fR, or \fIchild\fR
.TP
\fB-S\fR \fIattr\fR
sort results by attribute \fIattr\fR
.TP
\fB-w\fR \fIpasswd\fR
bind password used for simple bind
.TP
\fB-W\fR
prompt for bind password used in simple bind
.TP
\fB-x\fR
use simple authentication for bind
.TP
\fB-y\fR \fIfile\fR
read bind password from file
.TP
\fB-Y\fR \fImech\fR
SASL mechanism used during bind
.TP
\fB-z\fR \fIlimit\fR
size limit for search
.TP
\fB-Z\fR[\fB-Z\fR]
Issue  StartTLS before bind request. \fB-ZZ\fR requires TLS operations to be successful.
.TP
\fIfilter\fR
The search filter. If not provided, the default filter, \fB(objectclass=*)\fR,
is used.
.TP
\fIattributes...\fR
The list of \fIattributes\fR to display within the graph.

.SH EXAMPLE
The following command:
.in +4n
.nf

ldaptree -x -b o=internet --noleafs --compact

.fi
.in
will perform search on the LDAP server.  The output might look something the following:
.in +4n
.nf

#
# base: o=internet with scope subtree
# filter: (objectclass=*)
#
+--dc=net, o=internet
    +--dc=example
    |  +--ou=Groups
    |  \\--ou=People
    \\--dc=syzdek
       +--ou=Groups
       \\--ou=People

.fi
.in
The following command:
.in +4n
.nf

ldaptree -LLL -x -b o=internet '(objectclass=*)' givenname sn mail member description

.fi
.in

will perform search on the LDAP server.  The output might look something the following:
.in +4n
.nf

+--dc=net, o=internet
   +--dc=example
   |  +--ou=Groups
   |  |  +--cn=lug
   |  |  |     description: Linux Users' Group
   |  |  |     member: uid=dnullman,ou=People,dc=example,dc=net,o=internet
   |  |  |     member: uid=doughboy42,ou=People,dc=example,dc=net,o=internet
   |  |  |
   |  |  \\--cn=foodie
   |  |        description: People obsessed with good food
   |  |        member: uid=doughboy42,ou=People,dc=example,dc=net,o=internet
   |  |
   |  \\--ou=People
   |     +--uid=dnullman
   |     |     sn: Nullman
   |     |     givenName: Devian
   |     |     mail: noreply@example.com
   |     |
   |     \\--uid=jdough
   |           sn: Dough
   |           givenName: John
   |           mail: doughboy42@example.com
   |
   \\--dc=syzdek
      +--ou=Groups
      \\--ou=People
         \\--uid=syzdek
               sn: Syzdek
               givenName: David M.
               mail: david@syzdek.net

.fi
.in
The following command:
.in +4n
.nf

ldaptree -x -b o=internet --style=bullets --compact '(objectclass=*)' givenname sn mail member description

.fi
.in
will perform search on the LDAP server.  The output might look something the following:
.in +4n
.nf

* dc=net, o=internet
  * dc=example
    * ou=Groups
      * cn=lug
        * Attributes
          - description: Linux Users' Group
          - member: uid=dnullman,ou=People,dc=example,dc=net,o=internet
          - member: uid=doughboy42,ou=People,dc=example,dc=net,o=internet
      * cn=foodie
        * Attributes
          - description: People obsessed with good food
          - member: uid=doughboy42,ou=People,dc=example,dc=net,o=internet
    * ou=People
      * uid=dnullman
        * Attributes
          - sn: Nullman
          - givenName: Devian
          - mail: noreply@example.com
      * uid=jdough
          - sn: Dough
          - givenName: John
          - mail: doughboy42@example.com
  * dc=syzdek
    * ou=Groups
    * ou=People
      * uid=syzdek
        * Attributes
          - sn: Syzdek
          - givenName: David M.
          - mail: david@syzdek.net

.fi
.in

.SH "SEE ALSO"
.BR ldapsearch (1),
.BR ldap2csv (1),
.BR ldapdebug (1),
.BR ldap.conf (5)


.SH AUTHOR
David M. Syzdek <david@syzdek.net>


.SH ACKNOWLEDGEMENTS
\fB@PACKAGE_NAME@\fR is developed and maintained by David M. Syzdek
<david@syzdek.net>. \fB@PACKAGE_NAME@\fR utilizes and is styled after the
tools and libraries maintained by the \fBThe OpenLDAP Project\fR
<http://www.openldap.org/>.


.Sh CAVEATS

.\" end of man page


ChangeLog

LDAP Utilities
Copyright (C) 2012 Bindle Binaries <syzdek@bindlebinaries.com>.

0.3
      - refactoring libldaputils (syzdek)
      - adding TLS support

0.2   
      - Recreating Xcode project file. (syzdek)
      - Removing files installed by autotools from git repo. (syzdek)
      - Updating autotools files (Makefile.am, configure.ac, etc). (syzdek)
      - Changing versioning scheme. (syzdek)
      - Re-writing README. (syzdek)

0.1.1
      - Updating ldaputils_getpass() to use stderr.
      - Modifying behavior of --verbose to set debug level
      - Removing debugging lines for LDAP API Version.
      - Updating version information to include LDAP API information.
      - Adding sorting functions for LDAP searches.
      - Fixing problems compiling with MinGW.
      - Abstracting code for separating multivalue values.
      - Fixing gettext international support.
      - Changing how LDAP host/port/URI values are stored so that values
        from ldap.conf and ldaprc files will be used if they are not
        specified on the command line. (syzdek)
      - Changing default search scope to SUBTREE. (syzdek)
      - Changing version major version from 1.0.0. to 0.1.0 (syzdek)

0.1.0   2008/10/12
	* initial release of package (syzdek)


src/ldaptree.c
/*
 *  LDAP Utilities
 *  Copyright (C) 2019 David M. Syzdek <david@syzdek.net>.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are
 *  met:
 *
 *     1. Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *
 *     2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *
 *     3. Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived from
 *        this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 *  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 *  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 *  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 *  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
/**
 *  @file src/ldaptree.c display LDAP tree structure
 */
/*
 *  Simple Build:
 *     export CFLAGS='-DPROGRAM_NAME="ldaptree" -Wall -I../include'
 *     gcc ${CFLAGS} -c ldaptree.c
 *     gcc ${CFLAGS} -lldap -o ldaptree ldaptree.o ../lib/libldaputils.a
 *
 *  Libtool Build:
 *     export CFLAGS='-DPROGRAM_NAME="ldaptree" -Wall -I../include'
 *     libtool --mode=compile --tag=CC gcc ${CFLAGS} -c ldaptree.c
 *     libtool --mode=link    --tag=CC gcc ${CFLAGS} -lldap -o ldaptree \
 *             ldaptree.lo ../lib/libldaputils.a
 *
 *  Libtool Clean:
 *     libtool --mode=clean rm -f ldaptree.lo ldaptree
 */
#define _LDAP_UTILS_SRC_LDAP2CSV 1


///////////////
//           //
//  Headers  //
//           //
///////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Headers
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <getopt.h>
#include <assert.h>

#define LDAP_DEPRECATED 1
#include <ldap.h>
#include <ldaputils.h>


///////////////////
//               //
//  Definitions  //
//               //
///////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Definitions
#endif

#ifndef PROGRAM_NAME
#define PROGRAM_NAME "ldaptree"
#endif

#define MY_SHORT_OPTIONS LDAPUTILS_OPTIONS_COMMON LDAPUTILS_OPTIONS_SEARCH "87:6:5:4:3"


/////////////////
//             //
//  Datatypes  //
//             //
/////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Datatypes
#endif

/* configuration union */
typedef struct my_config MyConfig;
struct my_config
{
   LDAPUtils          * lud;
   int                  copy_entry;
   int                  pad0;
   char               * basedn;
   LDAPUtilsTreeOpts    treeopts;
};


//////////////////
//              //
//  Prototypes  //
//              //
//////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Prototypes
#endif

// main statement
int main(int argc, char * argv[]);

// parses configuration
int my_config(int argc, char * argv[], MyConfig ** cnfp);

// fress resources
void my_unbind(MyConfig * cnf);


/////////////////
//             //
//  Functions  //
//             //
/////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Functions
#endif

/// prints program usage and exits
void ldaputils_usage(void)
{
   printf("Usage: %s [options] [filter] [attributes...]\n", PROGRAM_NAME);
   ldaputils_usage_search(MY_SHORT_OPTIONS);
   ldaputils_usage_common(MY_SHORT_OPTIONS);
   printf("Display Options:\n");
   printf("  --noleaf                  do not print leaf nodes\n");
   printf("  --max-depth=num           maximum depth to display\n");
   printf("  --max-leafs=num           maximum leafs on each node to display\n");
   printf("  --max-nodes=num           maximum branches and leafs on each node to display\n");
   printf("  --style=format            output format of bullets or hierarchy (default: hierarchy)\n");
   printf("  --compact                 remove white space used for styling\n");
   printf("  --expand                  expand DN prefix\n");
   printf("\nReport bugs to <%s>.\n", PACKAGE_BUGREPORT);
   return;
}


/// main statement
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
int main(int argc, char * argv[])
{
   int                    err;
   int                    i;
   char                 * str;
   MyConfig             * cnf;
   LDAPMessage          * res;
   LDAPUtilsTree        * tree;

   cnf = NULL;

   // initializes resources and parses CLI arguments
   if ((err = my_config(argc, argv, &cnf)) != 0)
      return(1);
   if (!(cnf))
      return(0);

   // starts TLS and binds to LDAP
   if ((err = ldaputils_bind_s(cnf->lud)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldap_sasl_bind_s(): %s\n", cnf->lud->prog_name, ldap_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // performs LDAP search
   if ((err = ldaputils_search(cnf->lud, &res)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldaputils_search(): %s\n", cnf->lud->prog_name, ldap_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // retrieve entries
   if ((tree = ldaputils_get_tree(cnf->lud->ld, res, cnf->copy_entry)) == NULL)
   {
      fprintf(stderr, "%s: ldaputils_get_entries(): out of virtual memory\n", cnf->lud->prog_name);
      my_unbind(cnf);
      ldap_msgfree(res);
      return(1);
   };
   ldap_msgfree(res);

   // print header
   if (cnf->lud->silent < 2)
   {
      printf("#\n");
      ldap_get_option(cnf->lud->ld, LDAP_OPT_DEFBASE, &str);
      switch(cnf->lud->scope)
      {
         case LDAP_SCOPE_ONE:      printf("# base: %s with scope one\n", str); break;
         case LDAP_SCOPE_SUBTREE:  printf("# base: %s with scope subtree\n", str); break;
         case LDAP_SCOPE_BASE:     printf("# base: %s with scope base\n", str); break;
         case LDAP_SCOPE_CHILDREN: printf("# base: %s with scope children\n", str); break;
         default:                  printf("# base: %s\n", str); break;
      };
      printf("# filter: %s\n", cnf->lud->filter);
      if ( ((cnf->lud->attrs)) && ((cnf->copy_entry)) )
      {
         printf("# requesting:");
         for(i = 0; ((cnf->lud->attrs[i])); i++)
            printf(" %s", cnf->lud->attrs[i]);
         printf("\n");
      };
      printf("#\n");
   };

   // displays entries
   ldaputils_tree_print(tree, &cnf->treeopts);

   // frees resources
   ldaputils_tree_free(tree);
   my_unbind(cnf);

   return(0);
}


/// parses configuration
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
/// @param[in] cnfp   reference to configuration pointer
int my_config(int argc, char * argv[], MyConfig ** cnfp)
{
   int        c;
   int        err;
   int        option_index;
   MyConfig * cnf;

   static char   short_options[] = MY_SHORT_OPTIONS;
   static struct option long_options[] =
   {
      {"expand",         no_argument,      0, '2'},
      {"compact",        no_argument,      0, '3'},
      {"style",         required_argument, 0, '4'},
      {"style",         required_argument, 0, '4'},
      {"max-nodes",     required_argument, 0, '5'},
      {"maxnodes",      required_argument, 0, '5'},
      {"max-leafs",     required_argument, 0, '6'},
      {"maxleafs",      required_argument, 0, '6'},
      {"max-depth",     required_argument, 0, '7'},
      {"maxdepth",      required_argument, 0, '7'},
      {"no-leafs",      no_argument,       0, '8'},
      {"noleafs",       no_argument,       0, '8'},
      {"help",          no_argument,       0, 'h'},
      {"verbose",       no_argument,       0, 'v'},
      {"version",       no_argument,       0, 'V'},
      {NULL,            0,                 0, 0  }
   };

   // allocates memory for configuration
   if (!(cnf = (MyConfig *) malloc(sizeof(MyConfig))))
   {
      fprintf(stderr, "%s: out of virtual memory\n", PROGRAM_NAME);
      return(1);
   };
   memset(cnf, 0, sizeof(MyConfig));

   // initialize ldap utilities
   if ((err = ldaputils_initialize(&cnf->lud, PROGRAM_NAME)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldaputils_initialize(): %s\n", PROGRAM_NAME, ldap_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // loops through args
   option_index = 0;
   while((c = getopt_long(argc, argv, short_options, long_options, &option_index)) != -1)
   {
      switch(ldaputils_getopt(cnf->lud, c, optarg))
      {
         // shared option exit without error
         case -2:
         my_unbind(cnf);
         return(0);

         // no more arguments
         case -1:
         break;

         // long options toggles
         case 0:
         break;

         // shared option error
         case 1:
         my_unbind(cnf);
         return(1);

         case '2':
         cnf->treeopts.expandall = 1;
         break;

         case '3':
         cnf->treeopts.compact = 1;
         break;

         case '4':
         if (!(strcasecmp(optarg, "bullets")))
            cnf->treeopts.style = LDAPUTILS_TREE_BULLETS;
         else if (!(strcasecmp(optarg, "bullet")))
            cnf->treeopts.style = LDAPUTILS_TREE_BULLETS;
         else if (!(strcasecmp(optarg, "hierarchy")))
            cnf->treeopts.style = LDAPUTILS_TREE_HIERARCHY;
         else
         {
            fprintf(stderr, "%s: unrecognized style `--%s'\n", PROGRAM_NAME, optarg);
            fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
            my_unbind(cnf);
            return(1);
         };
         break;

         case '5':
         cnf->treeopts.maxchildren = (size_t)atoll(optarg);
         break;

         case '6':
         cnf->treeopts.maxleafs = (size_t)atoll(optarg);
         break;

         case '7':
         cnf->treeopts.maxdepth = (size_t)atoll(optarg);
         break;

         case '8':
         cnf->treeopts.noleaf = 1;
         break;

         // argument error
         case '?':
         fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);

         // unknown argument error
         default:
         fprintf(stderr, "%s: unrecognized option `--%c'\n", PROGRAM_NAME, c);
         fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);
      };
   };

   // saves filter
   cnf->lud->filter = "(objectclass=*)";
   if (argc > (optind))
   {
      if ((index(argv[optind], '=')) != NULL)
      {
         cnf->lud->filter = argv[optind];
         optind++;
      };
   };

   // configures LDAP attributes to return in results
   if (argc > optind)
   {
      cnf->copy_entry = 1;
      if (!(cnf->lud->attrs = (char **) malloc(sizeof(char *) * (size_t)(argc-optind+1))))
      {
         fprintf(stderr, "%s: out of virtual memory\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);
      };
      for(c = 0; c < (argc-optind); c++)
         cnf->lud->attrs[c] = argv[optind+c];
      cnf->lud->attrs[c] = NULL;
   } else {
      if (!(cnf->lud->attrs = (char **) malloc(sizeof(char *) * 2)))
      {
         fprintf(stderr, "%s: out of virtual memory\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);
      };
      cnf->lud->attrs[0] = NULL;
      cnf->lud->attrs[1] = NULL;
      if ((cnf->lud->attrs[0] = strdup("structuralObjectClass")) == NULL)
      {
         fprintf(stderr, "%s: out of virtual memory\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);
      };
   };

   // reads password
   if ((err = ldaputils_pass(cnf->lud)) != 0)
   {
      my_unbind(cnf);
      return(1);
   };

   *cnfp = cnf;

   return(0);
}


// fress resources
void my_unbind(MyConfig * cnf)
{
   assert(cnf != NULL);

   if ((cnf->lud))
      ldaputils_unbind(cnf->lud);

   free(cnf);

   return;
}

/* end of source file */
